### ISSUE TITLE
Bug: Parameters custom provider's client is expected as a AWS SDK Client

### ISSUE THREAD

**Issue by karthikeyanjp** [aws-powertools/powertools-lambda-typescript#1813]
### Expected Behaviour

Custom Providers can have integrations outside AWS Managed services. They don't have to use AWS SDK V3 clients.

### Current Behaviour

Custom Providers' clients are expected to align with AWS SDK V3 client signatures.

### Code snippet

main.ts

```typescript
import { BaseProvider } from '@aws-lambda-powertools/parameters/base';

// Example Provider Client
class MyProviderClient {
  public constructor(public config: any){
  }
}

// Example Provider (Dummy implemenation)
class MyCustomProvider extends BaseProvider{
         
          public _get(name: string, options?: unknown): Promise<unknown> {
                    throw new Error('Method not implemented.');
          }
         
          public _getMultiple(path: string, options?: unknown): Promise<void | Record<string, unknown>> {
                    throw new Error('Method not implemented.');
          }

}

// Below instantiation shouldn't throw any error
const myprovider = new MyCustomProvider({
          proto:  MyProviderClient as new (config?: unknown) => MyProviderClient,
});
```

### Steps to Reproduce

1. Initiate a npm project and install `@aws-lambda-powertools/parameters@1.17.0`
2. Run `ts-node main.ts`
3. See Error

```
Failed to add user agent middleware Error: The client provided does not match the expected interface
    at addUserAgentMiddleware (/Users/c6x7sq/work/code/platform/oss/bugs/powertools-parameters/node_modules/@aws-lambda-powertools/commons/lib/awsSdk/userAgentMiddleware.js:52:19)
    at new BaseProvider (/Users/c6x7sq/work/code/platform/oss/bugs/powertools-parameters/node_modules/@aws-lambda-powertools/parameters/lib/base/BaseProvider.js:46:46)
    at new MyCustomProvider (/Users/c6x7sq/work/code/platform/oss/bugs/powertools-parameters/customprovider.ts:10:1)
    at Object.<anonymous> (/Users/c6x7sq/work/code/platform/oss/bugs/powertools-parameters/customprovider.ts:22:20)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Module.m._compile (/Users/c6x7sq/.nvm/versions/node/v18.15.0/lib/node_modules/ts-node/src/index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/c6x7sq/.nvm/versions/node/v18.15.0/lib/node_modules/ts-node/src/index.ts:1621:12)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Function.Module._load (node:internal/modules/cjs/loader:958:12)
```

### Possible Solution

Refer [code](https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/packages/parameters/src/base/BaseProvider.ts#L67)

Suspect the issue is `addUserAgentMiddleware` being called for both awsSDKV3Client and non-AWS clients. 

```typescript
    if (awsSdkV3Client) {
      if (!isSdkClient(awsSdkV3Client)) {
        console.warn(
          'awsSdkV3Client is not an AWS SDK v3 client, using default client'
        );
        this.client = new proto(clientConfig ?? {});
      } else {
        this.client = awsSdkV3Client;
      }
    } else {
      this.client = new proto(clientConfig ?? {});
    }
    addUserAgentMiddleware(this.client, 'parameters');
  }

``` 

Possible Solution:

```typescript
    if (awsSdkV3Client) {
      if (!isSdkClient(awsSdkV3Client)) {
        console.warn(
          'awsSdkV3Client is not an AWS SDK v3 client, using default client'
        );
        this.client = new proto(clientConfig ?? {});
      } else {
        this.client = awsSdkV3Client;
        addUserAgentMiddleware(this.client, 'parameters');
      }
    } else {
      this.client = new proto(clientConfig ?? {});
    }
  }

``` 

### Powertools for AWS Lambda (TypeScript) version

latest

### AWS Lambda function runtime

20.x

### Packaging format used

npm

### Execution logs

```Shell
Failed to add user agent middleware Error: The client provided does not match the expected interface
    at addUserAgentMiddleware (/Users/c6x7sq/work/code/platform/oss/bugs/powertools-parameters/node_modules/@aws-lambda-powertools/commons/lib/awsSdk/userAgentMiddleware.js:52:19)
    at new BaseProvider (/Users/c6x7sq/work/code/platform/oss/bugs/powertools-parameters/node_modules/@aws-lambda-powertools/parameters/lib/base/BaseProvider.js:46:46)
    at new MyCustomProvider (/Users/c6x7sq/work/code/platform/oss/bugs/powertools-parameters/customprovider.ts:10:1)
    at Object.<anonymous> (/Users/c6x7sq/work/code/platform/oss/bugs/powertools-parameters/customprovider.ts:22:20)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Module.m._compile (/Users/c6x7sq/.nvm/versions/node/v18.15.0/lib/node_modules/ts-node/src/index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/c6x7sq/.nvm/versions/node/v18.15.0/lib/node_modules/ts-node/src/index.ts:1621:12)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Function.Module._load (node:internal/modules/cjs/loader:958:12)
```
```

**Comment by am29d** [comment]
Hey @karthikeyanjp , 

this is indeed a bug that should not happen if you pass your custom client that is not AWS SDK. I will fix the issue quickly to unblock you.

**Comment by am29d** [comment]
Hey @karthikeyanjp ,

I had a hard time to reproduce the error in our tests and now I see the problem. The error you see and the stackTrace is a `console.warn` message in `addUserAgentMiddleware`, your code executes successfully. We throw an error if we can't add user agent but also catch it for a warning log. We don't want to block execution if there are any issues with the user agent, thus we just print it. 

While we could potentially remove our throw Error, it's still helpful to have a log inside the catch to understand what went wrong. 

With this in mind, you should be able to implement your custom provider for parameters. If there are any open questions let me know, otherwise I will close the issue.
